File: backend/app/models/llm.py
"""LLM provider configuration models."""
from __future__ import annotations

from enum import Enum
from typing import Optional

from pydantic import BaseModel


class LLMProvider(str, Enum):
    OPENAI = "openai"
    ANTHROPIC = "anthropic"
    AZURE_OPENAI = "azure"
    GOOGLE_GEMINI = "gemini"
    OLLAMA = "ollama"
    OPENAI_COMPATIBLE = "openai_compatible"


class ModelConfig(BaseModel):
    provider: LLMProvider
    model_id: str
    display_name: str
    max_context_tokens: int
    supports_tools: bool = True
    supports_streaming: bool = True
    supports_vision: bool = False


# Catalogue of supported models per provider
SUPPORTED_MODELS: list[ModelConfig] = [
    # OpenAI
    ModelConfig(provider=LLMProvider.OPENAI, model_id="gpt-4o", display_name="GPT-4o",
                max_context_tokens=128000, supports_vision=True),
    ModelConfig(provider=LLMProvider.OPENAI, model_id="gpt-4o-mini",
                display_name="GPT-4o Mini", max_context_tokens=128000),
    ModelConfig(provider=LLMProvider.OPENAI, model_id="gpt-4-turbo",
                display_name="GPT-4 Turbo", max_context_tokens=128000, supports_vision=True),

    # Anthropic
    ModelConfig(provider=LLMProvider.ANTHROPIC, model_id="claude-opus-4-6",
                display_name="Claude Opus 4.6", max_context_tokens=200000),
    ModelConfig(provider=LLMProvider.ANTHROPIC, model_id="claude-sonnet-4-6",
                display_name="Claude Sonnet 4.6", max_context_tokens=200000),
    ModelConfig(provider=LLMProvider.ANTHROPIC, model_id="claude-haiku-4-5-20251001",
                display_name="Claude Haiku 4.5", max_context_tokens=200000),

    # Google Gemini
    ModelConfig(provider=LLMProvider.GOOGLE_GEMINI, model_id="gemini-1.5-pro",
                display_name="Gemini 1.5 Pro", max_context_tokens=1000000, supports_vision=True),
    ModelConfig(provider=LLMProvider.GOOGLE_GEMINI, model_id="gemini-1.5-flash",
                display_name="Gemini 1.5 Flash", max_context_tokens=1000000),
]


class ProviderStatus(BaseModel):
    provider: LLMProvider
    available: bool
    models: list[str]
    error: Optional[str] = None

File: frontend/src/types/llm.ts
export type LLMProviderType =
  | 'openai'
  | 'anthropic'
  | 'azure'
  | 'gemini'
  | 'ollama'
  | 'openai_compatible'

export interface LLMModel {
  id: string
  displayName: string
  provider: LLMProviderType
  maxTokens: number
  supportsTools: boolean
  supportsStreaming: boolean
}

export interface LLMConfig {
  provider: LLMProviderType
  model: string
  temperature: number
  maxTokens: number
  systemPrompt?: string
}

export const PROVIDER_MODELS: Record<LLMProviderType, LLMModel[]> = {
  openai: [
    { id: 'gpt-4o', displayName: 'GPT-4o', provider: 'openai', maxTokens: 128000, supportsTools: true, supportsStreaming: true },
    { id: 'gpt-4o-mini', displayName: 'GPT-4o Mini', provider: 'openai', maxTokens: 128000, supportsTools: true, supportsStreaming: true },
    { id: 'gpt-4-turbo', displayName: 'GPT-4 Turbo', provider: 'openai', maxTokens: 128000, supportsTools: true, supportsStreaming: true },
  ],
  anthropic: [
    { id: 'claude-opus-4-6', displayName: 'Claude Opus 4.6', provider: 'anthropic', maxTokens: 200000, supportsTools: true, supportsStreaming: true },
    { id: 'claude-sonnet-4-6', displayName: 'Claude Sonnet 4.6', provider: 'anthropic', maxTokens: 200000, supportsTools: true, supportsStreaming: true },
    { id: 'claude-haiku-4-5-20251001', displayName: 'Claude Haiku 4.5', provider: 'anthropic', maxTokens: 200000, supportsTools: true, supportsStreaming: true },
  ],
  azure: [
    { id: 'gpt-4o', displayName: 'Azure GPT-4o', provider: 'azure', maxTokens: 128000, supportsTools: true, supportsStreaming: true },
  ],
  gemini: [
    { id: 'gemini-1.5-pro', displayName: 'Gemini 1.5 Pro', provider: 'gemini', maxTokens: 1000000, supportsTools: true, supportsStreaming: true },
    { id: 'gemini-1.5-flash', displayName: 'Gemini 1.5 Flash', provider: 'gemini', maxTokens: 1000000, supportsTools: true, supportsStreaming: true },
  ],
  ollama: [
    { id: 'llama3.1', displayName: 'Llama 3.1', provider: 'ollama', maxTokens: 128000, supportsTools: false, supportsStreaming: true },
    { id: 'mistral', displayName: 'Mistral', provider: 'ollama', maxTokens: 32000, supportsTools: false, supportsStreaming: true },
    { id: 'codellama', displayName: 'Code Llama', provider: 'ollama', maxTokens: 16000, supportsTools: false, supportsStreaming: true },
  ],
  openai_compatible: [
    { id: 'custom', displayName: 'Custom Model', provider: 'openai_compatible', maxTokens: 32000, supportsTools: true, supportsStreaming: true },
  ],
}

export const PROVIDER_DISPLAY_NAMES: Record<LLMProviderType, string> = {
  openai: 'OpenAI',
  anthropic: 'Anthropic',
  azure: 'Azure OpenAI',
  gemini: 'Google Gemini',
  ollama: 'Ollama (Local)',
  openai_compatible: 'OpenAI-Compatible',
}

export const PROVIDER_COLORS: Record<LLMProviderType, string> = {
  openai: 'bg-green-600',
  anthropic: 'bg-orange-600',
  azure: 'bg-blue-600',
  gemini: 'bg-purple-600',
  ollama: 'bg-gray-600',
  openai_compatible: 'bg-cyan-600',
}
